engines/scumm/saveload.cpp

```diff
@@ -178,6 +175,323 @@ static bool saveSaveGameHeader(Common::WriteStream *out, SaveGameHeader &hdr) {
+			screen = vs->getPixels(0, _screenTop + i);
+			memcpy(&tempBitmap[_screenWidth * i], screen, _screenWidth * sizeof(byte));
+		}
+
+		for (int i = 0; i < 256; i++) {
+			_savegameThumbnailPalette[i] = getPackedRGBColorFromPalette(_currentPalette, i);
+		}
+
+		for (int i = 0; i < 120; i++) {
+			for (int j = 0; j < 160; j++) {
+				_savegameThumbnail[i * 160 + j] = tempBitmap[4 * (i * _screenWidth + j)];
+			}
+		}
+
+		free(tempBitmap);
+	}
+}
+
+bool ScummEngine_v8::fetchInternalSaveStateThumbnail(int slotId, bool isHeapSave) {
+	SaveGameHeader hdr;
+	Common::String filename;
+	Common::SeekableReadStream *in = openSaveFileForReading(slotId, isHeapSave, filename);
+	if (!in)
+		return false;
+
+	// In order to fetch the internal COMI thumbnail, we perform the same routine
+	// used during normal loading, stripped down to support only version 106 onwards...
+	if (!loadAndCheckSaveGameHeader(in, _game.heversion, hdr)) {
+		delete in;
+		return false;
+	}
+
+	if (hdr.ver > 0xFFFFFF)
+		hdr.ver = SWAP_BYTES_32(hdr.ver);
+
+	// Reject save games which do not contain the internal thumbnail...
+	if (hdr.ver < VER(106)) {
+		return false;
+	}
+
+	Graphics::skipThumbnail(*in);
+
+	SaveStateMetaInfos infos;
+	if (!loadInfos(in, &infos)) {
+		warning("Info section could not be found");
+		delete in;
+		return false;
+	}
+
+	hdr.name[sizeof(hdr.name) - 1] = 0;
+	_saveLoadDescription = hdr.name;
+
+
+	// Now do the actual loading
+	Common::Serializer ser(in, nullptr);
+	ser.setVersion(hdr.ver);
+	ser.syncArray(_savegameThumbnail, 19200, Common::Serializer::Byte, VER(106));
+	ser.syncArray(_savegameThumbnailPalette, 256, Common::Serializer::Uint32LE, VER(106));
+
+	delete in;
+	return true;
+}
+
+uint32 *ScummEngine_v8::fetchScummVMSaveStateThumbnail(int slotId, bool isHeapSave, int brightness) {
+	Common::String filename;
+	Graphics::Surface *thumbnailSurface;
+
+	// Perform the necessary steps to arrive at the thumbnail section of the save file...
+	Common::SeekableReadStream *in = openSaveFileForReading(slotId, isHeapSave, filename);
+	if (in) {
+		// We don't perform checks on the header: if we're here it means that the
+		// savestate follows the correct format and it is loadable.
+		in->skip(sizeof(uint32) * 3 + sizeof(SaveGameHeader::name));
+
+		// Load the thumbnail.
+		// We're under the assumption that its resolution will always be 160x120,
+		// which is a fourth of the original 640x480 internal resolution, so there's
+		// no need to scale the surface.
+		bool thumbSuccess = Graphics::loadThumbnail(*in, thumbnailSurface);
+		delete in;
+
+		if (thumbSuccess) {
+			// Now take the pixels from the surface, extract the RGB components, process them
+			// with the brightness parameter, and store them in an appropriate structure
+			// which the SCUMM graphics pipeline can use...
+			byte r, g, b;
+			uint32 *processedThumbnail = (uint32 *)malloc(sizeof(uint32) * thumbnailSurface->w * thumbnailSurface->h);
```

Comment:
using `new uint32[]` here lets you not test if the result is valid just below.
If you change it, don't forget to use `delete[]` everywhere the returned value is used.

Reply:
Thanks, great suggestion! That will make the code a little bit more slim, which is always appreciated :-)